File: browser/src/js/buyer/nexus.js
// Copyright (c) 2020 Jarret Dyrbye
// Distributed under the MIT software license, see the accompanying
// file LICENSE or http://www.opensource.org/licenses/mit-license.php

const Timestamp = require('../moneysocket/utl/timestamp.js').Timestamp;
const Nexus = require("../moneysocket/nexus/nexus.js").Nexus;

const RequestOpinionSeller = require(
    "./request_opinion_seller.js").RequestOpinionSeller;
const RequestOpinionInvoice = require(
    "./request_opinion_invoice.js").RequestOpinionInvoice;

const NotifyOpinionSeller = require(
    "../seller/notify_opinion_seller.js").NotifyOpinionSeller;
const NotifyOpinionSellerNotReady = require(
    "../seller/notify_opinion_seller_not_ready.js").NotifyOpinionSellerNotReady;
const NotifyOpinionInvoice = require(
    "../seller/notify_opinion_invoice.js").NotifyOpinionInvoice;
const NotifyOpinion = require("../seller/notify_opinion.js").NotifyOpinion;

const LAYER_NOTIFICATIONS = new Set(["NOTIFY_OPINION_SELLER",
                                     "NOTIFY_OPINION_SELLER_NOT_READY",
                                     "NOTIFY_OPINION_INVOICE",
                                     "NOTIFY_OPINION",
                                   ]);

class BuyerNexus extends Nexus {
    constructor(below_nexus, layer) {
        super(below_nexus, layer);

        this.onsellerinfo = null;
        this.onopinioninvoice = null;
        this.onopinion = null;

        this.consumerFinishedCb = null;
        this.request_reference_uuid = null;
        this.handshake_finished = false;
    }

    ///////////////////////////////////////////////////////////////////////////

    isLayerMessage(msg) {
        if (msg['message_class'] != "NOTIFICATION") {
            return false;
        }
        return LAYER_NOTIFICATIONS.has(msg['notification_name']);
    }

    onMessage(below_nexus, msg) {
        //console.log("consumer nexus got message");
        if (! this.isLayerMessage(msg)) {
            super.recvFromBelowCb(below_nexus, msg)
            return;
        }
        if (msg['notification_name'] == "NOTIFY_OPINION_SELLER") {
            if (! this.handshake_finished) {
                this.handshake_finished = true;
                this.sellerFinishedCb(this);
            }
            var seller_info = msg;
            if (this.onsellerinfo != null) {
                this.onsellerinfo(this, seller_info);
            }
        } else if (msg['notification_name'] == "NOTIFY_OPINION_INVOICE") {
            console.log("opinion invoice: " + JSON.stringify(msg));
            if (this.onopinioninvoice != null) {
                this.onopinioninvoice(this, msg['bolt11'],
                    msg['request_reference_uuid']);
            }
        } else if (msg['notification_name'] == "NOTIFY_OPINION") {
            if (this.onopinion != null) {
                this.onopinion(this, msg['item_id'], msg['opinion']);
            }
        }
    }

    onBinMessage(below_nexus, msg_bytes) {
        console.log("buyer nexus got raw msg from below");
    }

    startHandshake(sellerFinishedCb) {
        this.sellerFinishedCb = sellerFinishedCb;
        console.log("req seller");
        this.send(new RequestOpinionSeller());
    }

    requestOpinionInvoice(item_id) {
        console.log("req opinion invoice");
        this.send(new RequestOpinionInvoice(item_id));
    }
}

exports.BuyerNexus = BuyerNexus;

File: python/moneysocket/protocol/consumer/nexus.py
# Copyright (c) 2020 Jarret Dyrbye
# Distributed under the MIT software license, see the accompanying
# file LICENSE or http://www.opensource.org/licenses/mit-license.php

import uuid
import time
import logging

from twisted.internet.task import LoopingCall

from moneysocket.nexus.nexus import Nexus

from moneysocket.message.request.provider import RequestProvider
from moneysocket.message.request.ping import RequestPing


class ConsumerNexus(Nexus):
    def __init__(self, below_nexus, layer):
        super().__init__(below_nexus, layer)
        assert "notify_ping_cb" in dir(layer)
        self.consumer_finished_cb = None;
        self.handshake_finished = False;
        self.ping_interval = None;
        self.ping_start_time = None;


    ###########################################################################

    def is_layer_message(self, msg):
        if msg['message_class'] != "NOTIFICATION":
            return False
        return msg['notification_name'] in {"NOTIFY_PROVIDER",
                                            "NOTIFY_PROVIDER_NOT_READY",
                                            "NOTIFY_PING"}

    def recv_from_below_cb(self, below_nexus, msg):
        logging.info("consumer nexus got msg")

        if not self.is_layer_message(msg):
            super().recv_from_below_cb(below_nexus, msg)
            return

        if msg['notification_name'] == "NOTIFY_PROVIDER":
            if not self.handshake_finished:
                self.handshake_finished = True
                # nexus is announced to the transact layer
                # TODO, this message might be better to handle in the transact
                # layer instead. Refactor might be needed.
                self.consumer_finished_cb(self)
            # pass the message above for the transact layer to process
            super().recv_from_below_cb(below_nexus, msg)
        elif msg['notification_name'] == "NOTIFY_PROVIDER_NOT_READY":
            logging.info("provider not ready, waiting")
        elif msg['notification_name'] == "NOTIFY_PONG":
            if not self.ping_start_time:
                return
            msecs = (time.time() - self.ping_start_time) * 1000;
            self.layer.notify_ping_cb(self, round(msecs))
            self.ping_start_time = None;

    def recv_raw_from_below_cb(self, below_nexus, msg_bytes):
        logging.info("provider nexus got raw msg")
        super().recv_raw_from_below_cb(below_nexus, msg_bytes)

    ###########################################################################


    def start_handshake(self, consumer_finished_cb):
        self.consumer_finished_cb = consumer_finished_cb;
        self.send(RequestProvider());

    def send_ping(self):
        self.ping_start_time = time.time()
        self.send(RequestPing());


    def start_pinging(self):
        logging.info("START PING");
        self.ping_interval = LoopingCall(self.send_ping)
        self.ping_interval.start(3.0, now=True)

    def stop_pinging(self):
        logging.info("STOP PING");
        if not self.ping_interval:
            return
        self.ping_interval.stop()
        self.ping_interval = None;
